lib/Conversion/ArithExtToArith/ArithExtToArith.cpp

```diff
@@ -13,6 +14,20 @@ namespace arith_ext {
 #define GEN_PASS_DEF_ARITHEXTTOARITH
 #include "lib/Conversion/ArithExtToArith/ArithExtToArith.h.inc"
 
+template <typename ValueOrOpResult>
+TypedAttr modulusHelper(IntegerAttr mod, ValueOrOpResult op, bool mul = false) {
+  auto width = getElementTypeOrSelf(op).getIntOrFloatBitWidth();
+  auto modWidth = (mod.getValue() - 1).getActiveBits();
+  width = std::max(width, mul ? 2 * modWidth : modWidth + 1);
+  auto intType = IntegerType::get(op.getContext(), width);
+  auto truncmod = mod.getValue().sextOrTrunc(width);
```

Comment:
Hmm I think I might need some help, but should we assume this will be an unsigned modulus and zext it instead? FWIW (and I think I'm not totally up to speed on the issues before), that a signed modulus like 2**32 - 1 will be treated like a -1 when doing a signed remsi operation. But the operations below use remui, so I'm unclear if it will end up treating the sign extended modulus as unsigned..

Reply:
Agree with zext.

Reply:
Thanks, good catch, this should not be signed!